src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -14143,10 +14150,26 @@ private SeparatedSyntaxList<TNode> ParseCommaSeparatedSyntaxList<TNode>(
 
                         if (shouldParseSeparatorOrElement())
                         {
-                            // If we got a semicolon instead of comma, consume it with error and act as if it were a comma.
-                            nodes.AddSeparator(this.CurrentToken.Kind == SyntaxKind.SemicolonToken
-                                ? this.EatTokenWithPrejudice(separatorTokenKind)
-                                : this.EatToken(separatorTokenKind));
+                            if (this.CurrentToken.Kind == SyntaxKind.SemicolonToken)
+                            {
+                                // If we got a semicolon instead of comma, consume it with error and act as if it were a
+                                // comma. Note: while odd, there are no strict requirements on the syntax kinds of
+                                // separators in separated syntax lists.  So it is acceptable and expected by consumers
+                                // that they may get semicolons or commas here (despite what the grammar may say).
+                                var semicolonToken = this.EatToken(SyntaxKind.SemicolonToken);
+                                nodes.AddSeparator(WithAdditionalDiagnostics(
+                                    semicolonToken,
+                                    this.GetExpectedTokenError(
+                                        SyntaxKind.CommaToken, SyntaxKind.SemicolonToken,
+                                        offset: semicolonToken.GetLeadingTriviaWidth(),
+                                        width: semicolonToken.Width)));
```

Comment:
inlined this, and specifically not calling EatTokenWithPrejudiceForSkippedTokenList (since we're not adding to a skipped token list).